lib/repositories/message.js

```diff
@@ -394,20 +412,89 @@ function createApi(table, provider, sqlGenerator, dependencies) {
                 });
             });
         });
+    },
+
+    /**
+     * Deletes all messages belonging to the specific mailbox
+     */
+    removeByMailbox: function remove(mailbox) {
+      dependencies.logger.trace('message.removeByMailbox called');
+
+      var query = table
+        .select(table.star())
+        .from(table)
+        .where(table['mailbox_id'].equals(mailbox.getId()))
+        .toQuery();
+      query = provider.forUpdate(query);
+
+      return provider.beginTransaction(true)
+        .then(function(transaction) {
+
+          return transaction.runQuery(query)
+            .then(function(result) {
+              var query = table
+                .delete()
+                .where(table['mailbox_id'].equals(mailbox.getId()))
+                .toQuery();
+
+              var messages = result.rows;
+
+              var x;
+              for (x in messages) {
+                messages[x] = convertFromStorage(messages[x]);
+              }
+
+              return transaction.runQuery(query)
+                .then(function() {
+                  return transaction.commit();
+                })
+                .then(function() {
+                  dependencies.logger.debug({messages: messages},
+                                            'Messages removed');
+                  return messages;
+                });
+            })
+            .catch(function(err) {
+              return transaction.rollback()
+                .finally(function() {
+                  throw new Error(err);
+                });
+            });
+        });
     }
   };
 
   /**
    * Returns a count of all messages for the given mailbox and folder.
    */
-  function count(mailbox, folder) {
+  function count(fields) {
     dependencies.logger.trace('message.count called');
 
-     var query = table
+    var where;
+
+    if ('mailbox' in fields) {
```

Comment:
This is way too complicated for a count method. I would revert this method and just call the expected API for calculating a count. Duplication of a consistent API is not necessarily bad; especially if abstracting out the calls ends up less readable than the original format.